src/expressions.md

```diff
@@ -639,45 +632,113 @@ while i < 10 {
-loops](#infinite-loops), [break expressions](#break-expressions), and
-[continue expressions](#continue-expressions) for more information.
-
-## `for` expressions
+### Iterator loops
 
 A `for` expression is a syntactic construct for looping over elements provided
-by an implementation of `std::iter::IntoIterator`.
+by an implementation of `std::iter::IntoIterator`. If the iterator yields a
+value, that value is given the specified name and the body of the loop is
+executed, then control returns to the head of the `for` loop. If the iterator
+is empty, the `for` expression completes.
 
 An example of a `for` loop over the contents of an array:
 
 ```rust
-# type Foo = i32;
-# fn bar(f: &Foo) { }
-# let a = 0;
-# let b = 0;
-# let c = 0;
-
-let v: &[Foo] = &[a, b, c];
+let v = &["apples", "cake", "coffee"];
 
-for e in v {
-    bar(e);
+for text in v {
+    println!("I like {}.", text);
 }
 ```
 
 An example of a for loop over a series of integers:
 
 ```rust
-# fn bar(b:usize) { }
-for i in 0..256 {
-    bar(i);
+let mut sum = 0;
+for n in 1..11 {
+    sum += n;
+}
+assert_eq!(sum, 55);
+```
+
+### Loop labels
+
+A loop expression may optionally have a _label_. The label is written as
+a lifetime preceding the loop expression, as in `'foo: loop { break 'foo; }`,
+`'bar: while false {}`, `'humbug: for _ in 0..0 {}`.
+If a label is present, then labeled `break` and `continue` expressions nested
+within this loop may exit out of this loop or return control to its head.
+See [break expressions](#break-expressions) and [continue
+expressions](#continue-expressions).
+
+### `break` expressions
+
+When `break` is encountered, execution of the associated loop body is
+immediately terminated, for example:
+
+```rust
+let mut last = 0;
+for x in 1..100 {
+    if x > 12 {
+        break;
+    }
+    last = x;
 }
+assert_eq!(last, 12);
+```
+
+A `break` expression is normally associated with the innermost `loop`, `for` or
+`while` loop enclosing the `break` expression, but a [label](#loop-labels) can
+be used to specify which enclosing loop is affected. Example:
+
+```rust
+'outer: loop {
+    while true {
+        break 'outer;
+    }
+}
+```
+
+A `break` expression is only permitted in the body of a loop, and has one of
+the forms `break`, `break 'label` or (see [break-with-value](#break-with-value))
+`break EXPR` or `break 'label EXPR`.
+
+### `continue` expressions
+
+When `continue` is encountered, the current iteration of the associated loop
+body is immediately terminated, returning control to the loop *head*. In
+the case of a `while` loop, the head is the conditional expression controlling
+the loop. In the case of a `for` loop, the head is the call-expression
+controlling the loop.
+
+Like `break`, `continue` is normally associated with the innermost enclosing
+loop, but `continue 'label` may be used to specify the loop affected.
+A `continue` expression is only permitted in the body of a loop.
+
+### Break-with-value
```

Comment:
This can be worded a bit better. Something like "`break` with return values" maybe?